RV-Monitor/Kernel_Safety_Claims_by_Runtime_Verification_Monitors.md

```diff
@@ -0,0 +1,542 @@
+ \
+As an example, if there is a safety requirement associated with ioctl() (e.g. ioctl shall be used to set the correct timeout in an external safety watchdog device), even if the operation is in principle very simple, the Kernel code providing the implementation must be qualified to make sure:
+
+
+
+1. The kernel code implementing the requirement is correct;
+2. There is no unintended functionality in the current implementation;
+3. There is no undetected interference from other Kernel code outside the current implementation.
+
+With respect to points 1. and 2. above we can classify Kernel code into two classes:
+
+
+
+1. Kernel code that functionally contributes to the allocated safety requirement
+2. Kernel code potentially invoked as part of the implementation but not contributing to the safety requirement (e.g. printk).
+
+With respect to point 3. above we can have:
+
+
+
+3. Unintended functionalities due to interferences from Kernel code outside the implementation associated with classes A. and B. above. 
+
+
+When it comes to claim that code as in A. and B. are able to meet point 1 and point 2, for a complex SW component we need to hierarchically break it down into SW units, for each of these do a safety analysis and derive CoUs and Derived Safety Requirements; this regardless of code being classified as A. or B above. This process already proved to be very intensive even considering an entire subsystem like a single SW Unit (e.g. [STPA(-like) inside the Kernel ](https://docs.google.com/document/u/0/d/1K_cQSS2KYDnJQ0B91Zvlxq9-35Cx8ntbXwVMQJ51rvY/edit)).
+
+Finally with respect to failure modes as in C it is well known that, as of today, there are no comprehensive architectural measures that can isolate Kernel implementing the allocated safety requirement from code outside the implementation associated with A. and B.
+
+
+### Which problems are solved by runtime verification monitors
+
+[Runtime Verification Monitors](https://docs.kernel.org/trace/rv/index.html) are mechanisms inside the Kernel that allow monitoring the Kernel itself to behave according to a predefined machine state diagram. \
+So if the Kernel misbehaves leading to a violation of points 1 or 2 above, the runtime verification monitor detects that the monitored Kernel code has violated the associated machine state diagram and accordingly raises an exception halting the execution or doing other custom remedial actions. \
+In this regard RV Monitors could be used not only to do control flow monitoring but also data integrity monitoring. In fact it may be possible to associate states with corresponding valid data sets that could be stored internally as part of the monitor specific data structures and, upon incoming events, it may be possible to check the correctness of the event specific data against the current state and associated data set.
```

Comment:
*corresponding constraints or conditions

Reply:
Ok I can reword as "[...] with corresponding constraints and conditions; data sets could be stored internally as part of the monitor specific data [...]"

Reply:
ok